functions/src/offers/functions.ts

```diff
@@ -0,0 +1,78 @@
+import { validateOrReject } from 'class-validator';
+import * as admin from 'firebase-admin';
+import { Change, EventContext } from 'firebase-functions/lib/cloud-functions';
+import { DocumentSnapshot } from 'firebase-functions/lib/providers/firestore';
+
+import { IOffer, Offer, OfferStatus } from '../models/offers';
+import { RequestStatus } from '../models/requests';
+
+admin.initializeApp();
+const db = admin.firestore();
+const messaging = admin.messaging();
+
+const queueStatusUpdateTriggers = async (change: Change<DocumentSnapshot>): Promise<void[]> => {
+  const offerBefore = change.before.exists ? (change.before.data() as Offer) : null;
+  const offerAfter = change.after.data() as Offer;
+
+  const operations: Promise<void>[] = [];
+
+  // A request has just been accepted -- Update request with new information (CAV, new status)
+  if (offerBefore?.status !== OfferStatus.accepted && offerAfter?.status === OfferStatus.accepted) {
+    // Update Request with new status and CAV.
+    operations.push(
+      offerAfter.requestRef.update({
+        cavUserRef: offerAfter.cavUserRef,
+        status: RequestStatus.ongoing,
+      }),
+    );
+    
+    messaging.send({
+        data:{
+            entity: 'offer',
```

Comment:
if possible we should send data for the following:
- Render a "toast" notification on the front end which includes maybe sending some information about the offer
- Compose a text message with enough info for the user to act on it.

Ideally there would be a pub/sub cloud function that allows anyone to add messages to that get converted into SMS messages to the client (if they have `text me updates`) turned on.

This is down the road, but we should likely put a little more detail into the message body for these purposes.

Reply:
I was thinking of that but then I thought it was better to fetch that data from the device once the notification arrives as I had given the details of the entity in a generic form.
but Yes we can avoid additional reads with this method.
I will make the change

Reply:
also regarding twilio, I was under the impression that twilio was supposed to be done only through cloud functions.
We will expose a cloud function with HTTP invocation and look for authentication of users in it.

Reply:
Yes twilio should only be done from cloud functions (serverside), but I think hooking a cloud function trigger to the pub/sub topic allows us to concentrate the notification logic into the pub/sub topic for each user. This way anyone in the larger product can notify a user via SMS if publish to their notification topic vs having an HTTP endpoint. Simultaneously it creates a singular place to check the users own notification preferences.

Reply:
okay. that makes sense.

Reply:
I am making the modification now

Reply:
@sabind I tried looking into the parameters. but for the offer, the only attribute that can be used by the user to understand what the offer is about is the `message` attribute.
However, this may not be very meaningful for notification as the `message` may be out of context without the details about the `request` for which the offer was made.
We do have meaningful attributes in the `request`. But if we need to send them, we would need to fetch the details of the `request` which is an additional read on the trigger.
If we are doing another additional read, I feel it's better to do it on the `client-side` so as to be able to show the `request` as well if the user were to click on the notification.
However, if your goal is to add it for the purpose of the SMS, then it makes sense and I can read from the `request` object as well

Reply:
I'd simply say: " You have new offer to an open request: https://reach4help....<request_id>" and include the message.

Reply:
okay I have made the change for this and have pushed it now